sycl/include/CL/sycl/intel/sub_group.hpp

```diff
@@ -25,69 +26,167 @@ template <typename T, access::address_space Space> class multi_ptr;
+  OCLTy Arg = detail::convertDataToType<T, OCLTy>(x);
+  OCLTy Ret = __spirv_GroupBroadcast(__spv::Scope::Subgroup, Arg,
+                                            local_id.get(0));
+  return detail::convertDataToType<OCLTy, T>(Ret);
 }
 
-template <typename T, __spv::GroupOperation O>
-static typename std::enable_if<
-    !detail::is_floating_point<T>::value && std::is_unsigned<T>::value, T>::type
-calc(T x, intel::minimum<T> op) {
-  return __spirv_GroupUMin(__spv::Scope::Subgroup, O, x);
+template <typename T> T shuffle(T x, id<1> local_id) {
+  using OCLTy = detail::ConvertToOpenCLType_t<T>;
+  OCLTy Arg = detail::convertDataToType<T, OCLTy>(x);
+  OCLTy Ret = __spirv_SubgroupShuffleINTEL(Arg, local_id.get(0));
+  return detail::convertDataToType<OCLTy, T>(Ret);
 }
 
-template <typename T, __spv::GroupOperation O>
-static typename std::enable_if<detail::is_floating_point<T>::value, T>::type
-calc(T x, intel::minimum<T> op) {
-  return __spirv_GroupFMin(__spv::Scope::Subgroup, O, x);
+template <typename T> T shuffle_down(T current, T next, uint32_t delta) {
+  using OCLTy = detail::ConvertToOpenCLType_t<T>;
+  OCLTy Arg1 = detail::convertDataToType<T, OCLTy>(current);
+  OCLTy Arg2 = detail::convertDataToType<T, OCLTy>(next);
+  OCLTy Ret = __spirv_SubgroupShuffleDownINTEL(Arg1, Arg2, delta);
+  return detail::convertDataToType<OCLTy, T>(Ret);
 }
 
-template <typename T, __spv::GroupOperation O>
-static typename std::enable_if<
-    !detail::is_floating_point<T>::value && std::is_signed<T>::value, T>::type
-calc(T x, intel::maximum<T> op) {
-  return __spirv_GroupSMax(__spv::Scope::Subgroup, O, x);
+template <typename T> T shuffle_up(T previous, T current, uint32_t delta) {
+  using OCLTy = detail::ConvertToOpenCLType_t<T>;
+  OCLTy Arg1 = detail::convertDataToType<T, OCLTy>(previous);
+  OCLTy Arg2 = detail::convertDataToType<T, OCLTy>(current);
+  OCLTy Ret = __spirv_SubgroupShuffleUpINTEL(Arg1, Arg2, delta);
+  return detail::convertDataToType<OCLTy, T>(Ret);
 }
 
-template <typename T, __spv::GroupOperation O>
-static typename std::enable_if<
-    !detail::is_floating_point<T>::value && std::is_unsigned<T>::value, T>::type
-calc(T x, intel::maximum<T> op) {
-  return __spirv_GroupUMax(__spv::Scope::Subgroup, O, x);
+template <typename T> T shuffle_xor(T x, id<1> local_id) {
+  using OCLTy = detail::ConvertToOpenCLType_t<T>;
+  OCLTy Arg = detail::convertDataToType<T, OCLTy>(x);
+  OCLTy Ret = __spirv_SubgroupShuffleXorINTEL(
+      Arg, static_cast<uint32_t>(local_id.get(0)));
+  return detail::convertDataToType<OCLTy, T>(Ret);
 }
 
-template <typename T, __spv::GroupOperation O>
-static typename std::enable_if<detail::is_floating_point<T>::value, T>::type
-calc(T x, intel::maximum<T> op) {
-  return __spirv_GroupFMax(__spv::Scope::Subgroup, O, x);
+// Selects 8-bit, 16-bit or 32-bit type depending on size of T. If T doesn't
+// maps to mentioned types, then void is returned
+template <typename T>
+using SelectBlockTy =
```

Comment:
Yes it can cause conflicts and we had the problem before. So this is a real concern. But not using a `detail` namespace isn't OK. Because that could confuse users which would see extra types being available. I think there is no problem in having a `detail` namespace inside the `sub_group` namespace. It seems we are avoiding this so far to have different `detail` namespaces so far. I'm not sure why. If we stick with a single `detail` namespace we just need to make sure there isn't anything conflicting declared.

Reply:
> Yes it can cause conflicts and we had the problem before. So this is a real concern. But not using a detail namespace isn't OK.

Actually, I'm using it, these stuff is within `cl::sycl::detail::sub_group` namespace.

Reply:
Sorry I didn't see that.